File: code/cpp/second-minimum-node-in-a-binary-tree.cc
// 3ms, 09/07/2017
// Time  : O(n), worst case
// Space : O(h), worst case
int find_greater(TreeNode* root, int val)
{
    if (!root)
        return val;
    if (root->val != val)
        return root->val;
    int left_val = find_greater(root->left, val);
    int right_val = find_greater(root->right, val);

    return (left_val == val) ? right_val : (right_val == val) ? left_val : min(left_val, right_val);
}

int findSecondMinimumValue(TreeNode* root)
{
    if (!root || !root->left)
        return -1;
    int sec_min = find_greater(root, root->val);
    return (sec_min == root->val) ? -1 : sec_min;
}

File: code/python/second-minimum-node-in-a-binary-tree.py
# 03/17/2018
# Time  : O(n)
# Space : O(h)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def findSecondMinimumValue(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def find_min_two(root):
            if not root:
                return float('inf'), float('inf')
            min1 =  [x for x in find_min_two(root.left)  if x > root.val]
            min1 += [x for x in find_min_two(root.right) if x > root.val]
            return root.val, min(min1)
        
        min0, min1 = find_min_two(root)
        
        return min1 if min1 != float('inf') else -1
